compiler/rustc_middle/src/mir/interpret/pointer.rs

```diff
@@ -143,37 +162,78 @@ impl From<AllocId> for Pointer {
     }
 }
 
-impl Pointer<()> {
+impl<Tag> From<Pointer<Tag>> for Pointer<Option<Tag>> {
     #[inline(always)]
-    pub fn new(alloc_id: AllocId, offset: Size) -> Self {
-        Pointer { alloc_id, offset, tag: () }
+    fn from(ptr: Pointer<Tag>) -> Self {
+        let (tag, offset) = ptr.into_parts();
+        Pointer::new(Some(tag), offset)
     }
+}
+
+impl<Tag> Pointer<Option<Tag>> {
+    pub fn into_pointer_or_addr(self) -> Result<Pointer<Tag>, Size> {
+        match self.provenance {
+            Some(tag) => Ok(Pointer::new(tag, self.offset)),
+            None => Err(self.offset),
+        }
+    }
+
+    #[inline(always)]
+    pub fn map_erase_for_fmt(self) -> Pointer<Option<AllocId>>
+    where
+        Tag: Provenance,
+    {
+        Pointer { offset: self.offset, provenance: self.provenance.map(Provenance::erase_for_fmt) }
+    }
+}
 
+impl<Tag> Pointer<Option<Tag>> {
     #[inline(always)]
-    pub fn with_tag<Tag>(self, tag: Tag) -> Pointer<Tag> {
-        Pointer::new_with_tag(self.alloc_id, self.offset, tag)
+    pub fn null() -> Self {
+        Pointer { provenance: None, offset: Size::ZERO }
     }
 }
 
 impl<'tcx, Tag> Pointer<Tag> {
     #[inline(always)]
-    pub fn new_with_tag(alloc_id: AllocId, offset: Size, tag: Tag) -> Self {
-        Pointer { alloc_id, offset, tag }
+    pub fn new(provenance: Tag, offset: Size) -> Self {
+        Pointer { provenance, offset }
+    }
+
+    /// Obtain the constituents of this pointer. Not that the meaning of the offset depends on the type `Tag`!
+    /// This function must only be used in the implementation of `Machine::ptr_get_alloc`,
+    /// and when a `Pointer` is taken apart to be stored efficiently in an `Allocation`.
+    #[inline(always)]
+    pub fn into_parts(self) -> (Tag, Size) {
+        (self.provenance, self.offset)
+    }
+
+    #[inline(always)]
+    pub fn erase_for_fmt(self) -> Pointer
+    where
+        Tag: Provenance,
+    {
+        // FIXME: This is wrong! `self.offset` might be an absolute address.
+        Pointer { offset: self.offset, provenance: self.provenance.erase_for_fmt() }
```

Comment:
I still need to figure out a solution to this problem: we can't just change the `Tag` type since that affects how `offset` must be interpreted! (As absolute address vs as relative to some `AllocId`.)

I'll look into where this is currently actually needed, maybe there is a better way. The backup plan is to return a `Pointer<Option<Tag>>` and use the `None` provenance to reflect absolute addresses, but that would be an unfortunate loss of information in Miri traces.

Reply:
I think I can avoid pretty much all of those uses (though I am not quite sure yet how to best factor that) -- except for the one in `impl Display for ImmTy`. I am actually somewhat doubtful about that implementation... do we really want it to display the *pointee* of references etc? For Miri and even during a CTFE execution this will not work at all since it looks up the data only in the global `tcx`! So I think my preferred solution here would be to not call `pretty_print_const_scalar` at all, which would elegantly remove the problem...

Reply:
oof, yea that's annoying. I don't remember why we dispatch to the pretty printer, so yea, just printing without tcx or ecx information seems best

Reply:
Ah I think it is to get nice printing of scalars for some of the UB messages. I will keep using the const-pretty-print code for `ScalarInt` but not use it for `Pointer`.

Reply:
Resolved by latest commit.